tests/conftest.py

```diff
@@ -90,11 +95,36 @@ async def stream_writer(
     return async_buffers[1]
 
 
+def is_nullable_entity_field(field: dataclasses.Field) -> bool:
+    if get_origin(field.type) is not UnionType:
+        return False
+    try:
+        a, b = get_args(field.type)
+    except ValueError:
+        return False
+    return (a is NoneType and dataclasses.is_dataclass(b)) or (
+        b is NoneType and dataclasses.is_dataclass(a)
+    )
+
+
+def map_nullable_entity_fields(obj: DataclassInstance) -> dict[str, bool]:
+    """Return map of KIP-893 nullable entity fields."""
+    return {
+        field.name: is_nullable_entity_field(field) for field in dataclasses.fields(obj)
+    }
+
+
 class JavaTester:
     class _Encoder(JSONEncoder):
         def default(self, o: Any) -> Any:
             if dataclasses.is_dataclass(o):
-                return self._replace_tzaware_nulls(dataclasses.asdict(o))
+                return self._replace_tzaware_nulls(
+                    {
+                        k: v
+                        for k, v in dataclasses.asdict(o).items()
+                        if (v is not None or not map_nullable_entity_fields(o)[k])
```

Comment:
The nullable entity needed to be omitted entirely rather than passed as `null` to make the test pass.